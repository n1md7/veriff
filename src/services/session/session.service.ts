import { UUID } from '../../types/response.type';
import { VeriffAPI } from '../../http-clients/veriff/http.client';
import { ContextMediaResponse } from './responses/context-media.response';
import { SessionMediaResponse } from './responses/session-media.response';
import { SessionDetailResponse } from './responses/session-detail.response';
import { MediaService } from '../media/media.service';

export class SessionService {
  private contextMedia: ContextMediaResponse[] = [];
  private sessionMedia: SessionMediaResponse[] = [];
  private sessionDetail: SessionDetailResponse | null = null;

  private media = {
    front: new MediaService(),
    back: new MediaService(),
  };

  private indexedSessionMedia: Map<UUID, SessionMediaResponse> = new Map();

  public constructor(private readonly httpClient: VeriffAPI, private readonly requestID: string) {}

  public async run() {
    return this.fetchAndValidate().then(() => {
      this.indexSessionMedia();
      this.processMediaData();
    });
  }

  public getSessionDetails() {
    return this.sessionDetail;
  }

  public getProcessedMedia() {
    return this.media;
  }

  private fetchData() {
    const sessionDetails = this.httpClient.fetchSessionDetails(this.requestID);
    const sessionMedia = this.httpClient.fetchSessionMedia(this.requestID);
    const mediaContext = this.httpClient.fetchMediaContext(this.requestID);

    return Promise.all([sessionDetails, sessionMedia, mediaContext]);
  }

  private async fetchAndValidate() {
    const [sessionDetails, sessionMedia, mediaContext] = await this.fetchData();

    SessionMediaResponse.validate(sessionMedia);
    ContextMediaResponse.validate(mediaContext);
    SessionDetailResponse.validate(sessionDetails);

    this.contextMedia = mediaContext.map(ContextMediaResponse.from);
    this.sessionMedia = sessionMedia.map(SessionMediaResponse.from);
    this.sessionDetail = SessionDetailResponse.from(sessionDetails);
  }

  // Works in O(n) Time
  private indexSessionMedia() {
    for (const media of this.sessionMedia) {
      this.indexedSessionMedia.set(media.id, media);
    }
  }

  // Works in O(n) Time
  private processMediaData() {
    for (const contextMedia of this.contextMedia) {
      const sessionMedia = this.indexedSessionMedia.get(contextMedia.mediaId);
      if (sessionMedia) {
        if (contextMedia.isValidContext(contextMedia.context)) {
          this.media[contextMedia.context].pushMedia({
            mediaId: contextMedia.mediaId,
            mimeType: sessionMedia.mimeType,
            probability: contextMedia.probability,
          });
        }
      }
    }
  }
}
