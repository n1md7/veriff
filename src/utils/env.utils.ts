import dotenv from 'dotenv';
dotenv.config();

export class EnvUtils {
  static getServerPort() {
    return Number(process.env.PORT) || 3000;
  }

  static getVeriffBaseUrl() {
    const url = process.env.VERIFF_API_BASE_URL;
    if (!url) throw new Error('"VERIFF_API_BASE_URL" is missing');

    return url;
  }

  static getVeriffSessionUUID() {
    const uuid = process.env.VERIFF_SESSION_UUID;
    if (!uuid) throw new Error('"VERIFF_SESSION_UUID" is missing');

    return uuid;
  }
}
