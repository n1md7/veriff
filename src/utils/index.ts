export * from './string.utils';
export * from './response.utils';
export * from './env.utils';
