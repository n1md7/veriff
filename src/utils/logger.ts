export class Logger {
  private static message(msg: unknown, ctx: string, type: string) {
    console.log(`${type} - ${process.pid} - [${Date.now()}] [${ctx}]: ${JSON.stringify(msg)}`);
  }

  static log(msg: unknown, ctx: string) {
    Logger.message(msg, ctx, 'LOG');
  }

  static error(msg: unknown, ctx: string) {
    Logger.message(msg, ctx, 'ERROR');
  }

  static warn(msg: unknown, ctx: string) {
    Logger.message(msg, ctx, 'WARN');
  }
}
