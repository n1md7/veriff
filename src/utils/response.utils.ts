export class ResponseUtils {
  static getResponse(code, data) {
    const random = Math.floor(Math.random() * 10) + 1;
    if (random > 8) {
      return [500, 'There is something wrong with the service'];
    }

    return [code, data];
  }
}
