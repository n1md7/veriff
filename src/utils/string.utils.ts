import { EnvUtils } from './env.utils';

export const findMatch = (haystack: string, needle = EnvUtils.getVeriffSessionUUID()) => {
  return haystack.includes(needle);
};
