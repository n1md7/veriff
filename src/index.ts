import express from 'express';

import { xRequestId } from './middlewares/x-request-id.middleware';
import { errorHandler } from './middlewares/error.handler.middleware';
import { session } from './routes/sessions.route';
import { Logger } from './utils/logger';
import { EnvUtils } from './utils';

const app = express();
const port = EnvUtils.getServerPort();

app.use(xRequestId());
app.use('/api/sessions', session);
app.use(errorHandler());

app.listen(port, '0.0.0.0', () => {
  Logger.log(`Service is listening at http://localhost:${port}`, 'Init');
});

export { app };
