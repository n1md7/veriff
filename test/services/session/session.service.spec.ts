import { VeriffAPI } from '../../../src/http-clients/veriff/http.client';
import { SessionService } from '../../../src/services/session/session.service';
import { EnvUtils, ResponseUtils } from '../../../src/utils';

describe('SessionService', () => {
  const sessionId = EnvUtils.getVeriffSessionUUID();
  beforeAll(() => {
    jest.spyOn(ResponseUtils, 'getResponse').mockImplementation((code, data) => [code, data]);
  });
  afterAll(() => jest.clearAllMocks());
  afterEach(() => jest.clearAllMocks());
  it('should accumulate and sort request data by context', async () => {
    const api = new VeriffAPI(sessionId);
    const service = new SessionService(api, 'req-id');
    await service.run();

    const { front, back } = service.getProcessedMedia();
    expect(service.getSessionDetails()).toMatchSnapshot();
    expect(front.getSortedMedia()).toMatchSnapshot();
    expect(back.getSortedMedia()).toMatchSnapshot();
  });
  describe('throw on fetch', () => {
    it('should throw on fetchSessionDetails', () => {
      const api = new VeriffAPI(sessionId);
      jest.spyOn(api, 'fetchSessionDetails').mockRejectedValueOnce([{ error: 'fetchSessionDetails rejected' }] as any);
      const service = new SessionService(api, 'req-id');
      return expect(service.run()).rejects.toMatchSnapshot();
    });
    it('should throw on fetchSessionMedia', () => {
      const api = new VeriffAPI(sessionId);
      jest.spyOn(api, 'fetchSessionMedia').mockRejectedValueOnce([{ error: 'fetchSessionMedia rejected' }] as any);
      const service = new SessionService(api, 'req-id');
      return expect(service.run()).rejects.toMatchSnapshot();
    });
    it('should throw on fetchMediaContext', () => {
      const api = new VeriffAPI(sessionId);
      jest.spyOn(api, 'fetchMediaContext').mockRejectedValueOnce([{ error: 'fetchMediaContext rejected' }] as any);
      const service = new SessionService(api, 'req-id');
      return expect(service.run()).rejects.toMatchSnapshot();
    });
  });
  describe('throw on validation', () => {
    it('should throw on fetchSessionDetails validation', () => {
      const api = new VeriffAPI(sessionId);
      jest.spyOn(api, 'fetchSessionDetails').mockReturnValueOnce([{ invalid: 'fetchSessionDetails data' }] as any);
      const service = new SessionService(api, 'req-id');
      return expect(service.run()).rejects.toMatchSnapshot();
    });
    it('should throw on fetchMediaContext validation', () => {
      const api = new VeriffAPI(sessionId);
      jest.spyOn(api, 'fetchMediaContext').mockReturnValueOnce([{ invalid: 'fetchMediaContext data' }] as any);
      const service = new SessionService(api, 'req-id');
      return expect(service.run()).rejects.toMatchSnapshot();
    });
    it('should throw on fetchSessionMedia validation', () => {
      const api = new VeriffAPI(sessionId);
      jest.spyOn(api, 'fetchSessionMedia').mockReturnValueOnce([{ invalid: 'fetchSessionMedia data' }] as any);
      const service = new SessionService(api, 'req-id');
      return expect(service.run()).rejects.toMatchSnapshot();
    });
  });
});
