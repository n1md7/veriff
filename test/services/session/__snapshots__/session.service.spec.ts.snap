// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionService should accumulate and sort request data by context 1`] = `
SessionDetailResponse {
  "id": "90d61876-b99a-443e-994c-ba882c8558b6",
  "status": "internal_manual_review",
}
`;

exports[`SessionService should accumulate and sort request data by context 2`] = `
Array [
  Object {
    "mediaId": "40851916-3e86-45cd-b8ce-0e948a8a7751",
    "mimeType": "image/png",
    "probability": 0.9264236,
  },
  Object {
    "mediaId": "7f2dcbd8-5b5f-4f1a-bfa4-016ddf4dd662",
    "mimeType": "image/png",
    "probability": 0.8734357,
  },
  Object {
    "mediaId": "40f1e462-6db8-4313-ace3-83e4f5619c56",
    "mimeType": "image/png",
    "probability": 0.2931033,
  },
]
`;

exports[`SessionService should accumulate and sort request data by context 3`] = `
Array [
  Object {
    "mediaId": "a6c90b4f-ddfc-49eb-89ad-05b7f1274f96",
    "mimeType": "image/png",
    "probability": 0.9739324,
  },
]
`;

exports[`SessionService throw on fetch should throw on fetchMediaContext 1`] = `
Array [
  Object {
    "error": "fetchMediaContext rejected",
  },
]
`;

exports[`SessionService throw on fetch should throw on fetchSessionDetails 1`] = `
Array [
  Object {
    "error": "fetchSessionDetails rejected",
  },
]
`;

exports[`SessionService throw on fetch should throw on fetchSessionMedia 1`] = `
Array [
  Object {
    "error": "fetchSessionMedia rejected",
  },
]
`;

exports[`SessionService throw on validation should throw on fetchMediaContext validation 1`] = `[InvalidMediaContextException: "[0].id" is required]`;

exports[`SessionService throw on validation should throw on fetchSessionDetails validation 1`] = `[InvalidSessionDetailsException: "value" must be of type object]`;

exports[`SessionService throw on validation should throw on fetchSessionMedia validation 1`] = `[InvalidSessionMediaException: "[0].id" is required]`;
